### DISCUSSION TITLE
How to adjust the viewport with changes of camera

### DISCUSSION THREAD

**Discussion by yicong123456** [google/filament#7210] [Q&A]
I am new to filament and wanna use filament for 3d visualization in android hmi, I need to change the position of the camera with in the changes of vehicle and after the changes, I cannot get the correct scene to render. May I know if it is the correct way to adjust the viewport according to the camera everytime the position of camera changes so that I can get the correct result? Or if there are any demos or examples similarly. Thanks!

**Comment by romainguy** [comment]
You shouldn't need to change the viewport after moving the camera. To move the camera you need to set a new transform on it using the `TransformManager`. It sounds like the transform you are using might be incorrect? Can you share code and/or screenshots?

**Reply by ywang218** [reply]
Hi, thanks so much for the help! I got the camrea position every frame from the backend and I attempt to render the vehicle in front of the camera every frame when I get the position of it in the scene. so I used:view.viewport = Viewport((centerX-halfWidth).toInt(), (centerY-spaceFront).toInt(), viewport.width, viewport.height); and I print the result of left and bottom of the viewport is: left: 146253.8, bottom: 2556791.2. I also print out the vertex data I rendered who is within the range. I attached the code below.

**Reply by romainguy** [reply]
You shouldn't need to change the viewport, only reposition the camera (or change its `lookAt()`). The viewport is defined in pixels, it's about the resolution you want to render at, you shouldn't use the object's center or anything like that.

**Reply by yicong123456** [reply]
Got it. thanks so much for the explanation

**Comment by ywang218** [comment]
Oh, thanks so much for the suggestion, I did not use transformManager and I used Camera.lookAt .Here is my code: 

I retrieve data from backend for the position of car and other information and I have a World class to handle and manage the RenderObj's render state, the RenderObj is a super class while the Cube is the subclass represent the vehicle:

The class FrameCallback is the place I handle render:

inner class FrameCallback : Choreographer.FrameCallback {
 .......
override fun doFrame(frameTimeNanos: Long) {
            choreographer.postFrameCallback(this)
            if (uiHelper.isReadyToRender) {
                updateWorld()
                if (renderer.beginFrame(swapChain!!, frameTimeNanos)) {
                    renderer.render(view)
                    renderer.endFrame()
                }
  }

       fun initData(vertices:  MutableList<Vertex>, renderObj: RenderableObject, vertexSize: Int) {
            renderObj.addVertexData(vertices, vertexSize)
            renderObj.addIndexData(vertices)
            renderObj.addMaterial(material)
        }

        fun setCar(vehicle: Cube, cubeVertices: MutableList<Vertex>, carX: Float, carY: Float, carZ: Float,halfLen: Float, halfWidth: Float, halfHeight: Float, vertexSize: Int) {
            cubeVertices.add(Vertex(carX-halfLen, carY-halfHeight, carZ-halfWidth, 0xffff0000.toInt()));  // 0xffff0000
            cubeVertices.add(Vertex(carX+halfLen, carY-halfHeight, carZ-halfWidth, 0xff0000ff.toInt()));
            cubeVertices.add(Vertex(carX+halfLen, carY+halfHeight, carZ-halfWidth, 0xff0000ff.toInt()));
            cubeVertices.add(Vertex(carX-halfLen, carY+halfHeight, carZ-halfWidth, 0xff0000ff.toInt()));
            cubeVertices.add(Vertex(carX-halfLen, carY-halfHeight, carZ+halfWidth, 0xff0000ff.toInt()));
            cubeVertices.add(Vertex(carX+halfLen, carY-halfHeight, carZ+halfWidth, 0xff0000ff.toInt()));
            cubeVertices.add(Vertex(carX+halfLen, carY+halfHeight, carZ+halfWidth, 0xff0000ff.toInt()));
            cubeVertices.add(Vertex(carX-halfLen, carY+halfHeight, carZ+halfWidth, 0xff0000ff.toInt()));
            initData(cubeVertices, vehicle, vertexSize)
        }

fun updateWorld() {
            var dir = FloatArray(3); 
            var camPos = FloatArray(3);
            camPos[0] = this.mainVehicle?.position?.x?.toFloat()!!;
            camPos[1] = this.mainVehicle?.position?.y?.toFloat()!!;
            camPos[2] = this.mainVehicle?.position?.z?.toFloat()!!;
            val input: FloatArray = camera.getForwardVector(dir);
            val dirDis = dir.map { it * 500f }
            val tarPos: FloatArray = camPos + dir;

            val entityManager = EntityManager.get()
            if(entityManager != null) {
                entityManager.destroy(renderable)
            }
            world.clear()

            val intSize = 4
            val floatSize = 4
            val vertexSize = 3 * floatSize + intSize
            var mainVehicleInstance = Cube();
            val cubeVertices3 = mutableListOf<Vertex>()
            var carX3 = mainVehicle?.position?.x;
            if (carX3 != null && carX3 < 0) {
                    carX3 = -1.0f * carX3!!;
            }
            var carY3 = mainVehicle?.position?.y;
            val carZ3 = mainVehicle?.position?.z;
            var len3 = 2.0f; var height3 = 4.0f; var width3 = 2.0f;
            var halfLen3 = len3/2; var halfHeight3 = height3/2; var halfWidth3 = width3/2;
            if (carX3 != null && carY3 != null && carZ3 != null) {
                   val out = FloatArray(3);
                   val carPosArr = floatArrayOf(carX3.toFloat(), carY3.toFloat(), carZ3.toFloat())
                   cameraController = CameraController(camera, view, carPosArr)    // I use this camereaController to control the camera before 
                                                                                                                                     // I ask this question
                   cameraController.updateViewPort()
                   cameraController.updateCameraPosition()

                   setCar(mainVehicleInstance, cubeVertices3, carX3.toFloat(), carY3.toFloat(), carZ3.toFloat(), halfLen3, halfWidth3, halfHeight3, vertexSize)
            };

            world.addRenderable(mainVehicleInstance)
            world.update();
  }
}

Here is my CameraController class(I used in wrong way I think):
class CameraController(private val camera: Camera, private val view: View, private val objectPosition: FloatArray) {
    private var centerX: Float = 0.0f;
    private var centerY: Float = 0.0f;
    private var centerZ: Float = 0.0f;

    fun updateViewPort() {
        var viewport = view.viewport;
        val spaceFront = 1.0f;
        centerX = objectPosition[0];
        centerY = objectPosition[1];
        centerZ = objectPosition[2];
        val halfWidth = viewport.width / 2.0f;
        val halfHeight = viewport.height / 2.0f
        val viewPortLeft = centerX - halfWidth;
        val viewPortRight = centerX + halfWidth;
        val viewPortTop = centerY + spaceFront
        val viewPortBottom = centerY - halfHeight
        view.viewport = Viewport((centerX-halfWidth).toInt(), (centerY-spaceFront).toInt(), viewport.width, viewport.height);   // I set the 
                                                                                                                                                                                                //viewport here
    }

    fun updateCameraPosition() {
        camera.lookAt(centerX.toDouble(), (centerY-9.0f).toDouble(), centerZ.toDouble(), centerX.toDouble(), (centerY).toDouble(), centerZ.toDouble(), 0.0, 0.0, 1.0)  // I update the camera parameter here
    }
}

Here is the World class that I control the render:

class World <T : RenderableObject>(
     private val engine: Engine,
     private val scene: Scene
) {
    private val renderables = mutableListOf<T>()

    fun addRenderable(renderable: T) {
        renderables.add(renderable)
    }

    fun getData(): MutableList<T> {
        return this.renderables
    }

    fun update() {
        for (renderable in renderables) {
            renderable.update(scene, engine)
        }
    }

    fun clear() {
        renderables.clear()
    }
}

Here is the abstract class of RenderObj(which can be find in the initData function) : 
open class RenderableObject {
    lateinit var material: Material
    var vertexData:  ByteBuffer? = null
    var indexData: ByteBuffer? = null
    var vertexCount by Delegates.notNull<Int>()
    var verticesData: MutableList<VisualizationFragment.FrameCallback.Vertex> = mutableListOf()
    var indexChanged: Boolean = false
    var dataChanged: Boolean = false
    var type: RenderableManager.PrimitiveType = RenderableManager.PrimitiveType.LINES

    fun ByteBuffer.put(v: VisualizationFragment.FrameCallback.Vertex): ByteBuffer {
        putFloat(v.x)
        putFloat(v.y)
        putFloat(v.z)
        putInt(v.color)
        return this
    }

    fun setRenderType(type: RenderableManager.PrimitiveType) {
        this.type = type
    }

    fun getRenderType(): RenderableManager.PrimitiveType {
        return this.type
    }

    open fun update(scene: Scene, engine: Engine) {
    }

    open fun update() {

    }

    fun addMaterial(material: Material) {
        this.material = material;
    }

    fun addVertexData(vertices: MutableList<VisualizationFragment.FrameCallback.Vertex>, vertexSize: Int) {
        verticesData = vertices
        vertexCount = vertices.size
        vertexData = ByteBuffer.allocate(vertexCount * vertexSize)
            .order(ByteOrder.nativeOrder())
        for(value in vertices) {
            vertexData!!.put(value);
        }
        vertexData!!.flip();
    }

    fun addVertexData(vertexData: ByteBuffer, vertexCount: Int) {
        this.vertexData = vertexData
    }

    open fun addIndexData(vertices: MutableList<VisualizationFragment.FrameCallback.Vertex>) {

    }

    fun addIndexData(indexData: ByteBuffer, indexCount: Int) {
        this.indexData = indexData
    }

    fun clear() {}
}
and the Cube which is a subclass of RenderableObject is where I add entity to the scene:
class Cube : RenderableObject() {
    private var renderable = 0
    private val floatSize = 4
    private val vertexSize = 3 * floatSize + 4//3 * 4 // 3 float data and 4 byte for each data
    private lateinit var vertexBuffer: VertexBuffer
    private lateinit var indexBuffer: IndexBuffer
    data class float3(var x: Float, var y: Float, var z: Float)

    override fun update(scene: Scene, engine: Engine) {
        Timber.i("cube draw, ")
        vertexBuffer = VertexBuffer.Builder()
            .bufferCount(1)
            .vertexCount(vertexCount)
            .attribute(VertexBuffer.VertexAttribute.POSITION, 0, VertexBuffer.AttributeType.FLOAT3, 0,             vertexSize)
            .attribute(VertexBuffer.VertexAttribute.COLOR,    0, VertexBuffer.AttributeType.UBYTE4, 3 * floatSize, vertexSize)
            .normalized(VertexBuffer.VertexAttribute.COLOR)
            .build(engine)

        vertexData?.let { vertexBuffer.setBufferAt(engine, 0, it) }

        indexBuffer = IndexBuffer.Builder()
            .indexCount(vertexCount)
            .bufferType(IndexBuffer.Builder.IndexType.USHORT)
            .build(engine)
        indexData?.let { indexBuffer.setBuffer(engine, it) }


        renderable = EntityManager.get().create()

        RenderableManager.Builder(1)
            // Overall bounding box of the renderable
            .boundingBox(Box(0.0f, 0.0f, 0.0f, 1.0f, 1.0f, 0.01f))
            // Sets the mesh data of the first primitive
            .geometry(0, RenderableManager.PrimitiveType.TRIANGLES, vertexBuffer, indexBuffer, 0, 6*6)
            // Sets the material of the first primitive
            .material(0, material.defaultInstance)
            .build(engine, renderable)

        // Add the entity to the scene to render it
        scene.addEntity(renderable)
    }

    override fun addIndexData(vertices: MutableList<VisualizationFragment.FrameCallback.Vertex>) {
        val shortSize = 2
        val indexes = shortArrayOf(
            0, 1, 2,  0, 3, 2,
            4, 5, 6,  4, 6, 7,
            0, 4, 1,  1, 4, 5,
            2, 6, 3,  3, 6, 7,
            0, 3, 4,  3, 4, 7,
            1, 2, 5,  2, 5, 6
        )
        vertexCount = indexes.size
        indexData = ByteBuffer.allocate(vertexCount * shortSize)
            .order(ByteOrder.nativeOrder())
        for(value in indexes) {
            indexData!!.putShort(value)
        }
        indexData!!.flip()
    }
}.
